src/garage/tf/algos/npo.py

```diff
@@ -214,17 +224,20 @@ def _build_inputs(self):
 
             # old policy distribution
             policy_old_dist_info_vars = {
-                k: tf.compat.v1.placeholder(
-                    tf.float32,
-                    shape=[None] * 2 + list(shape),
-                    name='policy_old_%s' % k)
+                k: tf.compat.v1.placeholder(tf.float32,
+                                            shape=[None] * 2 + list(shape),
+                                            name='policy_old_%s' % k)
                 for k, shape in policy_dist.dist_info_specs
             }
             policy_old_dist_info_vars_list = [
                 policy_old_dist_info_vars[k]
                 for k in policy_dist.dist_info_keys
             ]
 
+            if self.flatten_input:
```

Comment:
can you write out some comments here describing the shapes of the inputs and their flattened representation? 

how does this case contrast with the one immediately below it, which flattens the observation? why is the flattening done twice?

Reply:
why is this step necessary if you already flattened the input in `process_samples`?

Reply:
In `process_samples` the samples (training data / real values) are flattened, while the flattening here is for the placeholders.

Reply:
ah. i think it would be more clear if you did this at placeholder creation-time.

Reply:
ah -- this would be much more clear if it were placed near placeholder creation and generously commented.

is there a way we can make the placeholder this dimension in the first place, rather than post-processing it?

Reply:
I thought it would be more straight forward if we put the boolean flag in `akro.space.to_tf_placeholder()`?

Reply:
yes -- but i'm trying to avoid bumping the akro version number. let's see if this feature is needed in more than one algorithm first.

remember the goal here is to get models in ASAP. NPO is gonna get rewritten to use tf.distributions soon, and that might let us get rid of (at least one kind of) some of the flattening.

Reply:
How about we do minimal changes - pass the boolean flag to algo and do the checking in sampler by accessing `self.algo.is_flatten_input`, and flatten if the flag is on? We can move the checking back to algo when we rewrite later. Agree that the goal is to get models in ASAP.

Reply:
that change may be minimal but it introduces a bunch of technical debt (and now you need to test a new case in the sampler).

again, why can't the algorithm just have a flag which lets it decide whether to flatten the observations?

today:
1. sampler produces flattened observations
1.  algo unconditionally flattens observations in `process_samples`
1. `optimize_policy` expects flattened observations (e.g. placeholders are flat)

tomorrow:
1. samples produces unflattened observations
1.  algo conditionally flattens observations in `process_samples`
1. `optimize_policy` checks whether to expect flattened or unflattened observations and actions.

it seems you are stuck on step (3) of "tomorrow" complexity-wise. rather than adding a whole new function to flatten your placeholder, why not infer its shape directly from the space? you know the [batch, time] dimensions already anyway.

you can file an issue in akro to add a flattening option in the tensor creation API.

```python
if self.flatten_inputs:
    obs_var = tf.compat.v1.placeholder(
        name='obs',
        dtype=tf.float32,
        shape=[None, None, self.observation_space.flat_dim])
else:
    obs_var = self.observation_space.to_tf_placeholder(name='obs', batch_dims=2)
```

Reply:
It's a good idea, and we can avoid the new function `flatten_input_along_dim`. Will do.